File: 0745-prefix-and-suffix-search/0745-prefix-and-suffix-search.cpp
struct Node{
    unordered_map<char,Node*> links;
    int ind;

    Node(){
        ind=-1;
    }
};
class WordFilter {
private:
      Node* root;
public:
    void insert(string &key,int ind){
       Node* node=root;
       for(auto ch: key){
          if(!node->links[ch]){
            node->links[ch]=new Node();
          }
          node=node->links[ch];
          node->ind=ind;
       }
    }
    WordFilter(vector<string>& words) {
         root=new Node();
         for(int index=0;index<words.size();index++){
            for(int i=0;i<=words[index].size();i++){
                string suf=words[index].substr(i);
                string key=suf+"#"+words[index];
                insert(key,index);
            }
         }
    }
    
    int f(string pref, string suff) {
        string key=suff+"#"+pref;
        Node* node=root;
        for(auto ch: key){
            if(!node->links[ch]) return -1;
            node=node->links[ch];
        }
        return node->ind;
    }
};

/**
 * Your WordFilter object will be instantiated and called as such:
 * WordFilter* obj = new WordFilter(words);
 * int param_1 = obj->f(pref,suff);
 */

File: 0745-prefix-and-suffix-search/0745-prefix-and-suffix-search.java
class Node{
    HashMap<Character,Node> links=new HashMap<>();
    int ind=-1;
    
    Node(){}
};
class WordFilter {
    private Node root=new Node();
    private void insert(String word,int ind){
        Node node=root;
        for(char ch: word.toCharArray()){
            if(!node.links.containsKey(ch)){
                node.links.put(ch,new Node());
            }
            node=node.links.get(ch);
            node.ind=ind;
        }
    }
    private int search(String word){
        Node node=root;
        for(char ch: word.toCharArray()){
            node=node.links.get(ch);
            if(node==null) return -1;
        }
        return node.ind;
    }
    public WordFilter(String[] words) {
        for(int index=0;index<words.length;index++){
            for(int i=0;i<words[index].length();i++){
                String suff=words[index].substring(i);
                String key=suff+"#"+words[index];
                insert(key,index);
            }
        }
    }
    
    public int f(String pref, String suff) {
        String key=suff+"#"+pref;
        return search(key);
    }
}

/**
 * Your WordFilter object will be instantiated and called as such:
 * WordFilter obj = new WordFilter(words);
 * int param_1 = obj.f(pref,suff);
 */